File: src/lib/visualizer/floorPlane.ts
/**
 * Floor plane estimation for perspective-correct marble projection.
 *
 * Instead of full Hough Transform + Canny edges (complex, slow in browser),
 * we use the already-refined SAM floor mask and apply linear regression to
 * its left/right boundary rows.  For hallway / corridor / living-room shots
 * this gives accurate perspective lines without OpenCV.
 *
 * The output trapezoid is then:
 *   1. Rasterized and intersected with the SAM mask (removes wall noise).
 *   2. Used as the homography source quad for perspective tile projection.
 *
 * Three-pass detection in renderFloorTexture:
 *   Pass 1 — standard band (fitStartFrac=0.15, fitEndFrac=0.80)
 *   Pass 2 — wider band  (fitStartFrac=0.05, fitEndFrac=0.95)
 *   Pass 3 — floorBoundingBox fallback (always high-confidence, less accurate VP)
 */

import type { Pt, Quad } from "./perspective";

export type FloorPlaneResult = {
  trapezoid:      Quad;              // [TL, TR, BR, BL] in image coordinates
  vanishingPoint: Pt;                // where the left/right perspective lines meet
  horizonY:       number;            // y-coordinate of estimated horizon line
  confidence:     "high" | "low";   // "low" triggers the next pass
};

/**
 * Estimate the floor plane from a refined binary floor mask (1 = floor).
 *
 * Prerequisites:
 *   – Mask must already be clipped above 40% height (no ceiling).
 *   – Mask must already be BFS-connected to the bottom edge.
 *
 * @param fitStartFrac  Fraction from top of floor span to start regression (default 0.15).
 * @param fitEndFrac    Fraction from top of floor span to end regression   (default 0.80).
 *
 * Returns confidence="low" when:
 *   – Regression lines are nearly parallel AND row extents are degenerate.
 *   – Estimated vanishing point falls below the floor top (backwards perspective).
 *   – Fitted trapezoid is too narrow or too wide to be a real floor.
 */
export function estimateFloorTrapezoid(
  floorMask:   Uint8Array,
  W:           number,
  H:           number,
  fitStartFrac = 0.15,
  fitEndFrac   = 0.80,
): FloorPlaneResult {
  // ── 1. Build per-row left/right extents ───────────────────────────────────
  const rowLeft  = new Int32Array(H).fill(W);
  const rowRight = new Int32Array(H).fill(-1);
  let topRow = H, bottomRow = -1;

  for (let y = 0; y < H; y++) {
    for (let x = 0; x < W; x++) {
      if (floorMask[y * W + x] === 1) {
        if (x < rowLeft[y])  rowLeft[y]  = x;
        if (x > rowRight[y]) rowRight[y] = x;
        if (y < topRow)   topRow    = y;
        bottomRow = y;
      }
    }
  }

  const FALLBACK: FloorPlaneResult = {
    trapezoid:      [{ x: 0, y: 0 }, { x: W, y: 0 }, { x: W, y: H }, { x: 0, y: H }],
    vanishingPoint: { x: W / 2, y: H / 2 },
    horizonY:       H / 2,
    confidence:     "low",
  };

  if (bottomRow < 0 || topRow >= bottomRow) return FALLBACK;

  const floorSpan = bottomRow - topRow;
  if (floorSpan < 20) return FALLBACK;

  // ── 2. Sample the fitting band ────────────────────────────────────────────
  const fitStartY = Math.round(topRow + floorSpan * fitStartFrac);
  const fitEndY   = Math.round(topRow + floorSpan * fitEndFrac);

  const leftPts:  Array<[number, number]> = [];
  const rightPts: Array<[number, number]> = [];

  for (let y = fitStartY; y <= fitEndY; y++) {
    if (rowLeft[y]  < W)  leftPts.push([y, rowLeft[y]]);
    if (rowRight[y] >= 0) rightPts.push([y, rowRight[y]]);
  }

  if (leftPts.length < 5 || rightPts.length < 5) return FALLBACK;

  // ── 3. Linear regression: x = a*y + b ────────────────────────────────────
  const fitL = fitXofY(leftPts);
  const fitR = fitXofY(rightPts);

  // ── 4. Vanishing point = intersection of the two lines ───────────────────
  const dA = fitL.a - fitR.a;
  let vpY: number, vpX: number, confidence: "high" | "low";

  if (Math.abs(dA) < 0.005) {
    // Nearly parallel — head-on or low-angle camera.
    // Use the actual row extents at the far and near edges — this IS the correct
    // geometry for a near-orthographic view; no regression fitting needed.
    const tlX = clampX(rowLeft[topRow],     W);
    const trX = clampX(rowRight[topRow],    W);
    const blX = clampX(rowLeft[bottomRow],  W);
    const brX = clampX(rowRight[bottomRow], W);

    const nearW = Math.abs(brX - blX);
    const farW  = Math.abs(trX - tlX);

    if (nearW > W * 0.04 && farW > W * 0.04) {
      return {
        trapezoid: [
          { x: Math.min(tlX, trX), y: topRow    },
          { x: Math.max(tlX, trX), y: topRow    },
          { x: Math.max(blX, brX), y: bottomRow },
          { x: Math.min(blX, brX), y: bottomRow },
        ],
        vanishingPoint: { x: W / 2, y: topRow - floorSpan },
        horizonY:       topRow,
        confidence:     "high",
      };
    }
    return { ...FALLBACK, vanishingPoint: { x: W / 2, y: topRow }, horizonY: topRow };
  }

  vpY = (fitR.b - fitL.b) / dA;
  vpX = fitL.a * vpY + fitL.b;

  const vpAbove    = vpY < topRow;
  const vpCentered = vpX > W * 0.05 && vpX < W * 0.95;
  const farW  = Math.abs((fitR.a * topRow    + fitR.b) - (fitL.a * topRow    + fitL.b));
  const nearW = Math.abs((fitR.a * bottomRow + fitR.b) - (fitL.a * bottomRow + fitL.b));
  // Relaxed from 0.85 → 0.98: handles wide shallow rooms and gentle perspective
  const hasConvergence = nearW > 0 && farW / nearW < 0.98;

  confidence = vpAbove && vpCentered && hasConvergence ? "high" : "low";

  // ── 5. Evaluate fitted lines at the top/bottom rows ──────────────────────
  const tlX = clampX(Math.round(fitL.a * topRow    + fitL.b), W);
  const trX = clampX(Math.round(fitR.a * topRow    + fitR.b), W);
  const blX = clampX(Math.round(fitL.a * bottomRow + fitL.b), W);
  const brX = clampX(Math.round(fitR.a * bottomRow + fitR.b), W);

  const trapezoid: Quad = [
    { x: Math.min(tlX, trX), y: topRow    },
    { x: Math.max(tlX, trX), y: topRow    },
    { x: Math.max(blX, brX), y: bottomRow },
    { x: Math.min(blX, brX), y: bottomRow },
  ];

  // ── 6. Sanity-check trapezoid width ───────────────────────────────────────
  const trapW     = trapezoid[2].x - trapezoid[3].x;
  const trapWFrac = trapW / W;
  if (trapWFrac < 0.05 || trapWFrac > 0.99) {
    return { ...FALLBACK, vanishingPoint: { x: vpX, y: vpY }, horizonY: vpY };
  }

  return {
    trapezoid,
    vanishingPoint: { x: vpX, y: vpY },
    horizonY: Math.max(0, Math.min(H, vpY)),
    confidence,
  };
}

/**
 * Bounding-box fallback: return the per-row extents at the top and bottom of
 * the floor mask as a trapezoid.  Used as Pass 3 when both regression passes
 * return low confidence.  Geometry is always valid; perspective accuracy is
 * lower than regression-fitted lines but the render will never be blank.
 */
export function floorBoundingBox(
  floorMask: Uint8Array,
  W: number,
  H: number,
): FloorPlaneResult {
  const rowLeft  = new Int32Array(H).fill(W);
  const rowRight = new Int32Array(H).fill(-1);
  let topRow = H, bottomRow = -1;

  for (let y = 0; y < H; y++) {
    for (let x = 0; x < W; x++) {
      if (floorMask[y * W + x] === 1) {
        if (x < rowLeft[y])  rowLeft[y]  = x;
        if (x > rowRight[y]) rowRight[y] = x;
        if (y < topRow)   topRow    = y;
        bottomRow = y;
      }
    }
  }

  if (bottomRow < 0) {
    return {
      trapezoid:      [{ x: 0, y: 0 }, { x: W, y: 0 }, { x: W, y: H }, { x: 0, y: H }],
      vanishingPoint: { x: W / 2, y: H / 2 },
      horizonY:       H / 2,
      confidence:     "low",
    };
  }

  const tlX = rowLeft[topRow]     < W  ? rowLeft[topRow]     : 0;
  const trX = rowRight[topRow]    >= 0 ? rowRight[topRow]    : W;
  const blX = rowLeft[bottomRow]  < W  ? rowLeft[bottomRow]  : 0;
  const brX = rowRight[bottomRow] >= 0 ? rowRight[bottomRow] : W;

  return {
    trapezoid: [
      { x: Math.min(tlX, trX), y: topRow    },
      { x: Math.max(tlX, trX), y: topRow    },
      { x: Math.max(blX, brX), y: bottomRow },
      { x: Math.min(blX, brX), y: bottomRow },
    ],
    vanishingPoint: { x: W / 2, y: topRow },
    horizonY:       topRow,
    confidence:     "high",
  };
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

/** Least-squares fit of x = a * y + b from [y, x] point pairs. */
function fitXofY(points: Array<[number, number]>): { a: number; b: number } {
  const n = points.length;
  if (n < 2) return { a: 0, b: points[0]?.[1] ?? 0 };

  let sumY = 0, sumX = 0, sumYX = 0, sumY2 = 0;
  for (const [y, x] of points) {
    sumY += y; sumX += x; sumYX += y * x; sumY2 += y * y;
  }
  const denom = n * sumY2 - sumY * sumY;
  if (Math.abs(denom) < 1e-10) return { a: 0, b: sumX / n };
  const a = (n * sumYX - sumY * sumX) / denom;
  const b = (sumX - a * sumY) / n;
  return { a, b };
}

function clampX(x: number, W: number): number {
  return Math.max(0, Math.min(W - 1, x));
}

File: services/visualizer/app/pipeline/floor_plane.py
"""
Floor plane / trapezoid estimation — port of src/lib/visualizer/floorPlane.ts.

Linear regression over the refined floor mask's row extents, same three-pass
strategy as the TS renderer (standard band, wide band, bounding-box fallback).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

import numpy as np

Quad = tuple[tuple[float, float], tuple[float, float], tuple[float, float], tuple[float, float]]
Confidence = Literal["high", "low"]


@dataclass
class FloorPlaneResult:
    trapezoid: Quad
    vanishing_point: tuple[float, float]
    horizon_y: float
    confidence: Confidence


def _row_extents(mask: np.ndarray) -> tuple[np.ndarray, np.ndarray, int, int]:
    H, W = mask.shape
    has_floor = mask.any(axis=1)
    row_left = np.full(H, W, dtype=np.int32)
    row_right = np.full(H, -1, dtype=np.int32)

    rows_with_floor = np.where(has_floor)[0]
    if rows_with_floor.size == 0:
        return row_left, row_right, H, -1

    for y in rows_with_floor:
        xs = np.where(mask[y] == 1)[0]
        row_left[y] = xs[0]
        row_right[y] = xs[-1]

    return row_left, row_right, int(rows_with_floor[0]), int(rows_with_floor[-1])


def _fit_x_of_y(points: list[tuple[float, float]]) -> tuple[float, float]:
    n = len(points)
    if n < 2:
        return 0.0, (points[0][1] if points else 0.0)

    ys = np.array([p[0] for p in points], dtype=np.float64)
    xs = np.array([p[1] for p in points], dtype=np.float64)
    sum_y, sum_x = ys.sum(), xs.sum()
    sum_yx = (ys * xs).sum()
    sum_y2 = (ys * ys).sum()
    denom = n * sum_y2 - sum_y * sum_y
    if abs(denom) < 1e-10:
        return 0.0, sum_x / n
    a = (n * sum_yx - sum_y * sum_x) / denom
    b = (sum_x - a * sum_y) / n
    return a, b


def _clamp_x(x: float, W: int) -> float:
    return max(0.0, min(W - 1.0, x))


def _fallback(W: int, H: int) -> FloorPlaneResult:
    return FloorPlaneResult(
        trapezoid=((0, 0), (W, 0), (W, H), (0, H)),
        vanishing_point=(W / 2, H / 2),
        horizon_y=H / 2,
        confidence="low",
    )


def estimate_floor_trapezoid(
    floor_mask: np.ndarray,
    fit_start_frac: float = 0.15,
    fit_end_frac: float = 0.80,
) -> FloorPlaneResult:
    H, W = floor_mask.shape
    row_left, row_right, top_row, bottom_row = _row_extents(floor_mask)

    fallback = _fallback(W, H)
    if bottom_row < 0 or top_row >= bottom_row:
        return fallback

    floor_span = bottom_row - top_row
    if floor_span < 20:
        return fallback

    fit_start_y = round(top_row + floor_span * fit_start_frac)
    fit_end_y = round(top_row + floor_span * fit_end_frac)

    left_pts = [(y, float(row_left[y])) for y in range(fit_start_y, fit_end_y + 1) if row_left[y] < W]
    right_pts = [(y, float(row_right[y])) for y in range(fit_start_y, fit_end_y + 1) if row_right[y] >= 0]

    if len(left_pts) < 5 or len(right_pts) < 5:
        return fallback

    a_l, b_l = _fit_x_of_y(left_pts)
    a_r, b_r = _fit_x_of_y(right_pts)

    d_a = a_l - a_r
    if abs(d_a) < 0.005:
        tl_x = _clamp_x(row_left[top_row], W)
        tr_x = _clamp_x(row_right[top_row], W)
        bl_x = _clamp_x(row_left[bottom_row], W)
        br_x = _clamp_x(row_right[bottom_row], W)
        near_w = abs(br_x - bl_x)
        far_w = abs(tr_x - tl_x)
        if near_w > W * 0.04 and far_w > W * 0.04:
            return FloorPlaneResult(
                trapezoid=(
                    (min(tl_x, tr_x), top_row),
                    (max(tl_x, tr_x), top_row),
                    (max(bl_x, br_x), bottom_row),
                    (min(bl_x, br_x), bottom_row),
                ),
                vanishing_point=(W / 2, top_row - floor_span),
                horizon_y=top_row,
                confidence="high",
            )
        return FloorPlaneResult(
            trapezoid=fallback.trapezoid,
            vanishing_point=(W / 2, top_row),
            horizon_y=top_row,
            confidence="low",
        )

    vp_y = (b_r - b_l) / d_a
    vp_x = a_l * vp_y + b_l

    vp_above = vp_y < top_row
    vp_centered = W * 0.05 < vp_x < W * 0.95
    far_w = abs((a_r * top_row + b_r) - (a_l * top_row + b_l))
    near_w = abs((a_r * bottom_row + b_r) - (a_l * bottom_row + b_l))
    has_convergence = near_w > 0 and far_w / near_w < 0.98

    confidence: Confidence = "high" if (vp_above and vp_centered and has_convergence) else "low"

    tl_x = _clamp_x(round(a_l * top_row + b_l), W)
    tr_x = _clamp_x(round(a_r * top_row + b_r), W)
    bl_x = _clamp_x(round(a_l * bottom_row + b_l), W)
    br_x = _clamp_x(round(a_r * bottom_row + b_r), W)

    trapezoid: Quad = (
        (min(tl_x, tr_x), top_row),
        (max(tl_x, tr_x), top_row),
        (max(bl_x, br_x), bottom_row),
        (min(bl_x, br_x), bottom_row),
    )

    trap_w = trapezoid[2][0] - trapezoid[3][0]
    trap_w_frac = trap_w / W
    if trap_w_frac < 0.05 or trap_w_frac > 0.99:
        return FloorPlaneResult(
            trapezoid=fallback.trapezoid,
            vanishing_point=(vp_x, vp_y),
            horizon_y=vp_y,
            confidence="low",
        )

    return FloorPlaneResult(
        trapezoid=trapezoid,
        vanishing_point=(vp_x, vp_y),
        horizon_y=max(0.0, min(H, vp_y)),
        confidence=confidence,
    )


def floor_bounding_box(floor_mask: np.ndarray) -> FloorPlaneResult:
    H, W = floor_mask.shape
    row_left, row_right, top_row, bottom_row = _row_extents(floor_mask)

    if bottom_row < 0:
        return _fallback(W, H)

    tl_x = float(row_left[top_row]) if row_left[top_row] < W else 0.0
    tr_x = float(row_right[top_row]) if row_right[top_row] >= 0 else float(W)
    bl_x = float(row_left[bottom_row]) if row_left[bottom_row] < W else 0.0
    br_x = float(row_right[bottom_row]) if row_right[bottom_row] >= 0 else float(W)

    return FloorPlaneResult(
        trapezoid=(
            (min(tl_x, tr_x), top_row),
            (max(tl_x, tr_x), top_row),
            (max(bl_x, br_x), bottom_row),
            (min(bl_x, br_x), bottom_row),
        ),
        vanishing_point=(W / 2, top_row),
        horizon_y=top_row,
        confidence="high",
    )
